File: A Recursion/recursion.c
#include <stdio.h>

// Sum of firts n Natural numbers-----------------
int sum(int n){
  if(n == 1)
    return 1;
  
  return sum(n-1) + n;
}

// Factorial of n---------------------------------
int fact(int n){
  if(n == 0)
    return 1;

  return fact(n-1)*n;
}

//Power m^n---------------------------------------
int pow(int m, int n){
  if(n == 0)
    return 1;
  
  return pow(m, n-1) * m;
}

//Power m^n, Fast method--------------------------
int power(int m, int n){
  if(n == 0)
    return 1;
  
  if(n % 2 == 0)
    return power(m*m, n/2);
  else
    return m * power(m*m, (n-1)/2);
}

// nth term of Fibonacci Series------------------
int fib1(int n){
  if(n <= 1)
    return n;
  
  return fib1(n-2) + fib1(n-1);
}

//Nth term of Fibonacci Series - Memoization------
int F[10];
int fibm(int n){
  if(n <= 1){
    F[n] = n;
    return n;
  }
  
  if(F[n-2] == -1)
    F[n-2] = fibm(n-2);

  if(F[n-1] == -1)
    F[n-1] = fibm(n-1);

  return F[n-2] + F[n-1];

}



//main function-----------------------------------
int main(){
  
  int n = 8;
  // printf("Sum of first %d Natural Numbers: %d \n", n, sum(n));
  // printf("Fatorial of %d: %d\n", n, fact(n));

  // printf("Power: %d\n", pow(3, 4));
  // printf("Power Fast: %d\n", power(3, 4));

  // int r = fib1(6);
  // printf("nth term of Fibonacci Series: %d\n", r);

  // Fibonacci Memoization
  for(int i=0; i<10; i++)
    F[i] = -1;

  int r = fibm(4);
  printf("nth term of Fibonacci Series: %d\n", r);

}

File: A Recursion/recursion.rb
# Sum of first n Natural Numbers--------------------
def Rsum(n)
  return 1 if n == 1

  Rsum(n - 1) + n
end

n = 8
puts "Sum of  first #{n} Natural Number: #{Rsum(n)}"

#-------------------------------------------------------

# Factorial
def fact(n)
  return 1 if n.zero?

  fact(n - 1) * n
end

puts "Factorial: #{fact(6)}"
#-------------------------------------------------------

# Power m^n
def pow(m, n)
  return 1 if n.zero?

  pow(m, n - 1) * m
end

# Power m^n, Fast Method
def power(m, n)
  return 1 if n.zero?

  n.even? ? power(m * m, n / 2) : m * power(m * m, (n - 1) / 2)

end

puts "Power: #{power(2, 9)}"

#---------------------------------------------------------

# Fibonacci series

def fib(n)
  return n if n <= 1
  
  fib(n - 2) + fib(n - 1)
end

puts "nth term of fibonacci series: #{fib(6)}"

# Fibonacci series - memoization
@ft = []
def fibm(n)
  if n <= 1
    @ft[n] = n
    return n
  end

  @ft[n-2] = fibm(n-2) if @ft[n-2].nil?
  @ft[n-1] = fibm(n-1) if @ft[n-1].nil?

  @ft[n-2] + @ft[n-1]
end

# Fibonacci serise - momization
@cache = [0, 1]
def fibb(n)
  return @cache[n] if @cache[n]

  @cache[n] = fibb(n-2) + fibb(n-1)
end

puts "nth term of Fibonacci series: #{fibb(7)}"


